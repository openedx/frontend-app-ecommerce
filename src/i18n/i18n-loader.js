/**
 * For each locale we want to support, react-intl needs 1) the locale-data, which includes
 * information about how to format numbers, handle plurals, etc., and 2) the translations, as an
 * object holding message id / translated string pairs.  A locale string and the messages object are
 * passed into the IntlProvider element that wraps your element hierarchy.
 *
 * Note that react-intl has no way of checking if the translations you give it actually have
 * anything to do with the locale you pass it; it will happily use whatever messages object you pass
 * in.  However, if the locale data for the locale you passed into the IntlProvider was not
 * correctly installed with addLocaleData, all of your translations will fall back to the default
 * (in our case English), *even if you gave IntlProvider the correct messages object for that
 * locale*.
 */

import { addLocaleData } from 'react-intl';
import Cookies from 'universal-cookie';

import arLocale from 'react-intl/locale-data/ar';
import enLocale from 'react-intl/locale-data/en';
import es419Locale from 'react-intl/locale-data/es';
import frLocale from 'react-intl/locale-data/fr';
import zhcnLocale from 'react-intl/locale-data/zh';

import COUNTRIES, { langs as countryLangs } from 'i18n-iso-countries';
import LANGUAGES, { langs as languageLangs } from '@cospired/i18n-iso-languages';

import arMessages from './messages/ar.json';
// no need to import en messages-- they are in the defaultMessage field
import es419Messages from './messages/es_419.json';
import frMessages from './messages/fr.json';
import zhcnMessages from './messages/zh_CN.json';

addLocaleData([...arLocale, ...enLocale, ...es419Locale, ...frLocale, ...zhcnLocale]);

// TODO: When we start dynamically loading translations only for the current locale, change this.
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/ar.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/en.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/es.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/fr.json'));
COUNTRIES.registerLocale(require('i18n-iso-countries/langs/zh.json'));

// TODO: When we start dynamically loading translations only for the current locale, change this.
// TODO: Also note that Arabic (ar) and Chinese (zh) are missing here.  That's because they're
// not implemented in this library.  If you read this and it's been a while, go check and see
// if that's changed!
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/en.json'));
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/es.json'));
LANGUAGES.registerLocale(require('@cospired/i18n-iso-languages/langs/fr.json'));

const messages = { // current fallback strategy is to use the first two letters of the locale code
  ar: arMessages,
  es: es419Messages,
  fr: frMessages,
  zh: zhcnMessages,
};

const cookies = new Cookies();

const getTwoLetterLanguageCode = code => code.substr(0, 2);

// Get the locale by setting priority. Skip if we don't support that language.
const getLocale = (localeStr) => {
  // 1. Explicit application request
  if (localeStr && messages[localeStr] !== undefined) {
    return localeStr;
  }
  // 2. User setting in cookie
  const cookieLangPref = cookies.get(process.env.LANGUAGE_PREFERENCE_COOKIE_NAME);
  if (cookieLangPref && messages[getTwoLetterLanguageCode(cookieLangPref)] !== undefined) {
    return getTwoLetterLanguageCode(cookieLangPref);
  }
  // 3. Browser language (default)
  return getTwoLetterLanguageCode(window.navigator.language);
};

const getMessages = (locale = getLocale()) => messages[locale];

const rtlLocales = ['ar', 'he', 'fa', 'ur'];
const isRtl = locale => rtlLocales.includes(locale);

const handleRtl = () => {
  if (isRtl(getLocale())) {
    document.getElementsByTagName('html')[0].setAttribute('dir', 'rtl');
    document.styleSheets[0].disabled = true;
  } else {
    document.styleSheets[1].disabled = true;
  }
};

/**
 * Provides a lookup table of country IDs to country names for the current locale.
 */
const getCountryMessages = (locale) => {
  const finalLocale = countryLangs().includes(locale) ? locale : 'en';

  return COUNTRIES.getNames(finalLocale);
};

/**
 * Provides a lookup table of language IDs to language names for the current locale.
 */
const getLanguageMessages = (locale) => {
  const finalLocale = languageLangs().includes(locale) ? locale : 'en';

  return LANGUAGES.getNames(finalLocale);
};

const sortFunction = (a, b) => {
  // If localeCompare exists, use that.  (Not supported in some older browsers)
  if (typeof String.prototype.localeCompare === 'function') {
    return a[1].localeCompare(b[1], getLocale());
  }
  if (a[1] === b[1]) {
    return 0;
  }
  // Otherwise make a best effort.
  return a[1] > b[1] ? 1 : -1;
};

/**
 * Provides a list of countries represented as objects of the following shape:
 *
 * {
 *   key, // The ID of the country
 *   name // The localized name of the country
 * }
 *
 * The list is sorted alphabetically in the current locale.
 * This is useful for select dropdowns primarily.
 */
const getCountryList = (locale) => {
  const countryMessages = getCountryMessages(locale);
  return Object.entries(countryMessages)
    .sort(sortFunction)
    .map(([code, name]) => ({ code, name }));
};

/**
 * Provides a list of languages represented as objects of the following shape:
 *
 * {
 *   key, // The ID of the language
 *   name // The localized name of the language
 * }
 *
 * The list is sorted alphabetically in the current locale.
 * This is useful for select dropdowns primarily.
 */
const getLanguageList = (locale) => {
  const languageMessages = getLanguageMessages(locale);
  return Object.entries(languageMessages)
    .sort(sortFunction)
    .map(([code, name]) => ({ code, name }));
};

export {
  getCountryList,
  getCountryMessages,
  getLanguageList,
  getLanguageMessages,
  getLocale,
  getMessages,
  handleRtl,
  isRtl,
};
