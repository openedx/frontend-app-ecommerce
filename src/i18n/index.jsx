import arMessages from './messages/ar.json';
import caMessages from './messages/ca.json';
// no need to import en messages-- they are in the defaultMessage field
import es419Messages from './messages/es_419.json';
import frMessages from './messages/fr.json';
import zhcnMessages from './messages/zh_CN.json';
import heMessages from './messages/he.json';
import idMessages from './messages/id.json';
import kokrMessages from './messages/ko_kr.json';
import plMessages from './messages/pl.json';
import ptbrMessages from './messages/pt_br.json';
import ruMessages from './messages/ru.json';
import thMessages from './messages/th.json';
import ukMessages from './messages/uk.json';

const messages = {
  ar: arMessages,
  'es-419': es419Messages,
  fr: frMessages,
  'zh-cn': zhcnMessages,
  ca: caMessages,
  he: heMessages,
  id: idMessages,
  'ko-kr': kokrMessages,
  pl: plMessages,
  'pt-br': ptbrMessages,
  ru: ruMessages,
  th: thMessages,
  uk: ukMessages,
};

export default messages;
