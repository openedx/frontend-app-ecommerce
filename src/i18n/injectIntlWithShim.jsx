import React from 'react';
import { injectIntl, intlShape } from 'react-intl';
import { logError } from '@edx/frontend-logging';


const injectIntlWithShim = (WrappedComponent) => {
  class ShimmedIntlComponent extends React.Component {
    static propTypes = {
      intl: intlShape.isRequired,
    };

    constructor(props) {
      super(props);
      this.shimmedIntl = Object.create(this.props.intl, {
        formatMessage: {
          value: (definition) => {
            if (definition === undefined || definition.id === undefined) {
              const error = new Error('i18n error: An undefined message was supplied to intl.formatMessage.');
              if (process.env.NODE_ENV !== 'production') {
                console.error(error); // eslint-disable-line no-console
                return '!!! Missing message supplied to intl.formatMessage !!!';
              }
              logError(error);
              return ''; // Fail silent in production
            }
            return this.props.intl.formatMessage(definition);
          },
        },
      });
    }

    render() {
      return <WrappedComponent {...this.props} intl={this.shimmedIntl} />;
    }
  }

  return injectIntl(ShimmedIntlComponent);
};


export default injectIntlWithShim;
