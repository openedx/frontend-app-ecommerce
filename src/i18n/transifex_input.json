{
  "siteheader.links.courses": "Courses",
  "siteheader.links.programs": "Programs & Degrees",
  "siteheader.links.schools": "Schools & Partners",
  "siteheader.user.menu.dashboard": "Dashboard",
  "siteheader.user.menu.profile": "Profile",
  "siteheader.user.menu.account.settings": "Account",
  "siteheader.user.menu.logout": "Logout",
  "siteheader.user.menu.login": "Login",
  "siteheader.user.menu.register": "Sign Up",
  "error.unexpected.message": "An unexpected error occurred.",
  "error.notfound.message": "The page you're looking for is unavailable or there's an error in the URL. Please check the URL and try again.",
  "root.page": "Congratulations!  You have a new micro-frontend.",
  "ecommerce.order.history.page.heading": "Order History",
  "ecommerce.order.history.no.orders": "Orders you place with edX will appear here.",
  "ecommerce.order.history.loading.orders": "Loading orders...",
  "ecommerce.order.history.loading.error": "Error: {error}"
}