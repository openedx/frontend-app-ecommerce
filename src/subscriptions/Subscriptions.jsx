import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getConfig } from '@edx/frontend-platform';
import { sendTrackEvent } from '@edx/frontend-platform/analytics';
import { FormattedMessage } from '@edx/frontend-platform/i18n';
import {
  ActionRow,
  AlertModal,
  Button,
  Hyperlink,
  StatefulButton,
} from '@edx/paragon';
import { SpinnerSimple, Info, Launch } from '@edx/paragon/icons';

import SubscriptionCardsView from './SubscriptionCardsView';
import SubscriptionUpsell from './SubscriptionUpsell';

import { clearStripeError, fetchStripeCustomerPortalURL } from './actions';
import { subscriptionsSelector } from './selectors';

const Subscriptions = () => {
  const dispatch = useDispatch();
  const {
    subscriptions,
    stripeCustomerPortalURL,
    stripeError,
    stripeLoading,
  } = useSelector(subscriptionsSelector);
  const hasSubscriptions = subscriptions.length > 0;

  const handleManageSubscriptionsClick = () => {
    sendTrackEvent('edx.bi.user.subscription.order-page.manage.clicked');
    dispatch(fetchStripeCustomerPortalURL());
  };

  const handeAlertClose = () => {
    dispatch(clearStripeError());
  };

  useEffect(() => {
    if (stripeCustomerPortalURL) {
      window.open(stripeCustomerPortalURL, '_blank', 'noopener,noreferrer');
    }
  }, [stripeCustomerPortalURL]);

  const buttonLabel = (
    <FormattedMessage
      id="ecommerce.order.history.subscriptions.manage.button"
      defaultMessage="Manage my subscriptions"
      description="Button text for managing subscriptions."
    />
  );

  const renderSpinner = () => (
    <div className="icon-spin">{SpinnerSimple()}</div>
  );

  const renderEmpty = () => (
    <>
      <FormattedMessage
        id="ecommerce.order.history.subscriptions.subtitle.empty"
        defaultMessage="You do not have any active or previous subscriptions."
        description="Subtitle for subscriptions section when there are no subscriptions."
      >
        {(text) => <span className="text-dark-900">{text}</span>}
      </FormattedMessage>
      <SubscriptionUpsell />
    </>
  );

  const renderSubscriptions = () => (
    <>
      <div className="section flex-md-row align-items-start align-items-md-center justify-content-between">
        <FormattedMessage
          id="ecommerce.order.history.subscriptions.subtitle"
          defaultMessage="To view your receipts, change your payment method or cancel your subscription, click {buttonLabel}."
          description="Subtitle for subscriptions section."
          values={{
            buttonLabel: <i>{buttonLabel}</i>,
          }}
        >
          {(text) => <span className="text-dark-900">{text}</span>}
        </FormattedMessage>
        <StatefulButton
          size="sm"
          className="text-nowrap"
          labels={{ default: buttonLabel }}
          icons={{ default: undefined }}
          iconAfter={stripeLoading ? renderSpinner : Launch}
          state={stripeLoading ? 'pending' : 'default'}
          onClick={handleManageSubscriptionsClick}
        />
      </div>
      <SubscriptionCardsView subscriptions={subscriptions} />
      <AlertModal
        variant="danger"
        title="Something went wrong."
        icon={Info}
        isOpen={stripeError}
        onClose={handeAlertClose}
        footerNode={(
          <ActionRow>
            <Button variant="tertiary" onClick={handeAlertClose}>
              Dismiss
            </Button>
          </ActionRow>
        )}
      >
        <FormattedMessage
          tagName="p"
          id="ecommerce.order.history.subscriptions.stripe.error"
          defaultMessage="Refresh this page and try again. If this problem persists, {supportLink}."
          description="Error message when Stripe subscription information fails to load."
          values={{
            supportLink: (
              <Hyperlink
                destination={`${getConfig().SUPPORT_URL}/hc/requests/new`}
              >
                <FormattedMessage
                  id="ecommerce.order.history.support.fragment"
                  defaultMessage="contact support"
                  description="The support link as in 'please {contact support}'"
                />
              </Hyperlink>
            ),
          }}
        />
      </AlertModal>
    </>
  );

  return (
    <section>
      <FormattedMessage
        tagName="h2"
        id="ecommerce.order.history.subscriptions.heading"
        defaultMessage="Subscriptions"
        description="Heading for subscriptions section."
      />
      {hasSubscriptions ? renderSubscriptions() : renderEmpty()}
    </section>
  );
};

export default Subscriptions;
