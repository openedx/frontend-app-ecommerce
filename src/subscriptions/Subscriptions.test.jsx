/* eslint-disable global-require */
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import configureMockStore from 'redux-mock-store';

import Subscriptions from './Subscriptions';

const mockStore = configureMockStore();
const storeMocks = require('../store/__mocks__/mockStore');

describe('<Subscriptions />', () => {
  describe('Renders correctly in various states', () => {
    it('renders with subscriptions', () => {
      const tree = renderer
        .create(
          <IntlProvider locale="en">
            <Provider store={mockStore(storeMocks)}>
              <Subscriptions />
            </Provider>
          </IntlProvider>,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
