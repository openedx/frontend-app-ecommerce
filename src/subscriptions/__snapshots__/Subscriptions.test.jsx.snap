// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Subscriptions /> Renders correctly in various states renders with subscriptions 1`] = `
<section>
  <h2>
    Subscriptions
  </h2>
  <div
    className="section flex-md-row align-items-start align-items-md-center justify-content-between"
  >
    <span
      className="text-dark-900"
    >
      To view your receipts, change your payment method or cancel your subscription, click 
      <i>
        Manage my subscriptions
      </i>
      .
    </span>
    <button
      aria-disabled={false}
      aria-live="assertive"
      className="pgn__stateful-btn pgn__stateful-btn-state-default text-nowrap btn btn-primary btn-sm"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <span
        className="d-flex align-items-center justify-content-center"
      >
        <span>
          Manage my subscriptions
        </span>
      </span>
      <span
        className="pgn__icon pgn__icon__sm btn-icon-after"
      >
        <svg
          aria-hidden={true}
          fill="none"
          focusable={false}
          height={24}
          role="img"
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 19H5V5h7V3H3v18h18v-9h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
            fill="currentColor"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="section section-gap-lg"
  >
    <a
      className="pgn__hyperlink default-link standalone-link"
      href="http://localhost:18000/dashboard/programs/a87e5eac-3c93-45a1-a8e1-4c79ca8401c8"
      onClick={[Function]}
      target="_self"
    >
      <div
        className="bg-light-200 p-3 pgn__card pgn__card-light card"
        tabIndex={-1}
      >
        <div
          className="section flex-column-reverse flex-sm-row align-items-start align-items-sm-center"
        >
          <h3
            className="text-info-500 m-0"
          >
            Blockchain Fundamentals
          </h3>
          <span
            className="text-capitalize badge badge-light"
          >
            trial
          </span>
        </div>
        <p
          className="small text-gray-500 m-0"
        >
          University of California, Berkeley
        </p>
      </div>
    </a>
    <a
      className="pgn__hyperlink default-link standalone-link"
      href="http://localhost:18000/dashboard/programs/0c6e5fa2-96e8-40b2-9ebe-c8b0df2a3b22"
      onClick={[Function]}
      target="_self"
    >
      <div
        className="bg-light-200 p-3 pgn__card pgn__card-light card"
        tabIndex={-1}
      >
        <div
          className="section flex-column-reverse flex-sm-row align-items-start align-items-sm-center"
        >
          <h3
            className="text-info-500 m-0"
          >
            Critical Thinking
          </h3>
          <span
            className="text-capitalize badge badge-light"
          >
            active
          </span>
        </div>
        <p
          className="small text-gray-500 m-0"
        >
          Simmons University
        </p>
      </div>
    </a>
    <a
      className="pgn__hyperlink default-link standalone-link"
      href="http://localhost:18000/dashboard/programs/a87e5eac-3c93-45a1-a8e1-4c79ca8401c8"
      onClick={[Function]}
      target="_self"
    >
      <div
        className="bg-light-200 p-3 pgn__card pgn__card-light card"
        tabIndex={-1}
      >
        <div
          className="section flex-column-reverse flex-sm-row align-items-start align-items-sm-center"
        >
          <h3
            className="text-info-500 m-0"
          >
            Blockchain Fundamentals
          </h3>
          <span
            className="text-capitalize badge badge-light"
          >
            inactive
          </span>
        </div>
        <p
          className="small text-gray-500 m-0"
        >
          University of California, Berkeley
        </p>
      </div>
    </a>
    <a
      className="pgn__hyperlink default-link standalone-link"
      href="http://localhost:18000/dashboard/programs/0c6e5fa2-96e8-40b2-9ebe-c8b0df2a3b22"
      onClick={[Function]}
      target="_self"
    >
      <div
        className="bg-light-200 p-3 pgn__card pgn__card-light card"
        tabIndex={-1}
      >
        <div
          className="section flex-column-reverse flex-sm-row align-items-start align-items-sm-center"
        >
          <h3
            className="text-info-500 m-0"
          >
            Critical Thinking
          </h3>
          <span
            className="text-capitalize badge badge-light"
          >
            inactive
          </span>
        </div>
        <p
          className="small text-gray-500 m-0"
        >
          Simmons University
        </p>
      </div>
    </a>
  </div>
</section>
`;
